controllers/eventhandlers/targetgrouppolicy.go

```diff
@@ -0,0 +1,158 @@
+		if errors.IsNotFound(err) {
+			h.log.Debugw("TargetGroupPolicy is referring to non-existent service, skipping",
+				"policyName", policyName, "serviceName", svcName.String())
+		} else {
+			// Still gracefully skipping the event but errors other than NotFound are bad sign.
+			h.log.Warnw("Failed to query targetRef of TargetGroupPolicy",
+				"policyName", policyName, "serviceName", svcName.String(), "reason", err.Error())
+		}
+		return nil
+	}
+	h.log.Debugw("TargetGroupPolicy change on Service detected",
+		"policyName", policyName, "serviceName", svcName.String())
+
+	return svc
+}
+
+func (h *targetGroupPolicyEventHandler) MapToHTTPRoute(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+
+	svc := h.getTargetRef(obj)
+	if svc == nil {
+		return nil
+	}
+
+	routeList := &gateway_api.HTTPRouteList{}
+	h.client.List(context.TODO(), routeList)
+
+	for _, httpRoute := range routeList.Items {
+		if isServiceUsedByRoute(core.NewHTTPRoute(httpRoute), svc) {
+			routeName := types.NamespacedName{
+				Namespace: httpRoute.Namespace,
+				Name:      httpRoute.Name,
+			}
+			requests = append(requests, reconcile.Request{NamespacedName: routeName})
+			h.log.Infow("Service TargetGroupPolicy change triggering HTTPRoute update",
+				"serviceName", svc.Namespace+"/"+svc.Name, "routeName", routeName.String())
+		}
+	}
+	return requests
+}
+
+func (h *targetGroupPolicyEventHandler) MapToGRPCRoute(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+
+	svc := h.getTargetRef(obj)
+	if svc == nil {
+		return nil
+	}
+
+	routeList := &gateway_api_v1alpha2.GRPCRouteList{}
+	h.client.List(context.TODO(), routeList)
+
+	for _, grpcRoute := range routeList.Items {
+		if isServiceUsedByRoute(core.NewGRPCRoute(grpcRoute), svc) {
+			routeName := types.NamespacedName{
+				Namespace: grpcRoute.Namespace,
+				Name:      grpcRoute.Name,
+			}
+			requests = append(requests, reconcile.Request{NamespacedName: routeName})
+			h.log.Infow("Service TargetGroupPolicy change triggering GRPCRoute update",
+				"serviceName", svc.Namespace+"/"+svc.Name, "routeName", routeName.String())
+		}
+	}
+	return requests
+}
+
+func (h *targetGroupPolicyEventHandler) MapToServiceExport(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+
+	svc := h.getTargetRef(obj)
+	if svc == nil {
+		return nil
+	}
+	svcName := types.NamespacedName{
+		Namespace: svc.Namespace,
+		Name:      svc.Name,
+	}
+	svcExport := &mcs_api.ServiceExport{}
+	if err := h.client.Get(context.TODO(), svcName, svcExport); err != nil {
+		return nil
+	}
+	requests = append(requests, reconcile.Request{
+		NamespacedName: svcName,
+	})
+	return requests
+}
+
+func isServiceUsedByRoute(route core.Route, svc *corev1.Service) bool {
+	for _, rule := range route.Spec().Rules() {
+		for _, backendRef := range rule.BackendRefs() {
+			if backendRef.Kind() != nil && string(*backendRef.Kind()) != "Service" {
```

Comment:
Don't we require ```Kind()``` to be ```"Service"```? Here it can be ```nil``` and still match.

Reply:
per [gwapi spec](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.BackendObjectReference) - Kind defaults to Service when unspecified

Reply:
The wording is kind of ambiguous. Does it mean if it's not specified the value will be set to "Service", or does it mean if it's not specified it should be interepreted as "Service". I would expect the former, but I can't find code that says so. I tried applying some config that didn't have the field specified, and it *looks* like it's being set to "Service".

My interpretation I think is that it actually should never be nil. I think for the purposes of our code, we should ensure it exists and equals "Service"

Reply:
yup it seems you're right. it should never be nil as it is defined as a default value in CRD. Fixed the logic and test case.